### ISSUE TITLE
Possible to have a :required attribute?

### ISSUE THREAD

**Issue by srchulo** [tobyink/p5-moops#5]
I find the :ro attribute super useful. I often want to use :ro and have all fields be required, too. Would it be possible to add this as an attribute?

**Comment by tobyink** [comment]
This seems like a reasonable request.

Attributes could be made non-required by explicitly specifying `required => false` or by providing a default.

**Comment by tobyink** [comment]
That would be the simplest way to implement it, yes.

**Comment by tobyink** [comment]
Released a new MooseX::MungeHas.

**Comment by srchulo** [comment]
Wow! That was fast :)

**Comment by tobyink** [comment]
It was only a few lines of code. The test cases were more work than the implementation.

Moops just needs this now:

```perl
use v5.14;
use strict;
use warnings FATAL => 'all';
no warnings qw(void once uninitialized numeric);

package Moops::TraitFor::Keyword::req;

our $AUTHORITY = 'cpan:TOBYINK';
our $VERSION   = '0.036';

use Moo::Role;

around arguments_for_moosex_mungehas => sub {
	my $next = shift;
	my $self = shift;
	require MooseX::MungeHas;
	MooseX::MungeHas->VERSION('0.011');
	return ('always_required', $self->$next(@_));
};

1;
```

Once you've got that, you should be able to do:

```
use Moops;

class Foo :req {
	...;
}
```

I'll try to get it included in a release of Moops some time this month. If you wanted to write some test cases for it, that might help me get it released faster. [t/94trait-ro-rw-rwp.t](https://github.com/tobyink/p5-moops/blob/master/t/94trait-ro-rw-rwp.t) is pretty close to the kinds of tests you'd want to do.

**Comment by srchulo** [comment]
Thank you! I'm on vacation without my laptop now, but I will try to get to this soon to help you out :)

**Comment by srchulo** [comment]
Created pull request here:

https://github.com/tobyink/p5-moops/pull/7

Sorry-- I'm not sure if there's a better way to link theses.

**Comment by srchulo** [comment]
Also, if you don't mind, could you explain how

`return ('always_required', $self->$next(@_));`

Works? It's returning '`always_required`' and the values return by `$next`, but to what? And with multiple traits do the return values just stack up, and `$next` can be another `around` method for another trait? It looks like they're passed to `use MooseX::MungeHas qw()`.

**Comment by tobyink** [comment]
`Moops::Keyword::Role` line 51-52 is:

```perl
push @lines, "use MooseX::MungeHas qw(@{[ $self->arguments_for_moosex_mungehas ]});"
	if $using =~ /^Mo/;
```

And `Moops::Keyword::Role->arguments_for_moosex_mungehas` just provides a list of strings.

Traits like `:req` get applied to Moops::Keyword::Role, so can modify the list returned by `arguments_for_moosex_mungehas`, in this case adding another string to it.

**Comment by tobyink** [comment]
Yep. That's just generally how `around` works in Moose/Mouse/Moo. Multiple roles can be applied to a class and can all use `around` to modify the same method. (The order in which they are applied is not guaranteed, but in this case, the order doesn't matter anyway.)

**Comment by srchulo** [comment]
Awesome, thank you :) Please let me know if you'd like any changes on the pull request :)

**Comment by tobyink** [comment]
Looks good to me.
